File: Python/Valid Sudoku.py
from typing import List


class Solution:
    """
    Time:   O(n^3)
    Memory: O(1)
    """

    SIZE = 3
    EMPTY = '.'

    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for i in range(self.SIZE ** 2):
            for j in range(self.SIZE ** 2):
                if board[i][j] == self.EMPTY:
                    continue

                for k in range(self.SIZE ** 2):
                    if board[i][k] == board[i][j] and k != j:
                        return False
                    if board[k][j] == board[i][j] and k != i:
                        return False

                row = i - i % self.SIZE
                col = j - j % self.SIZE
                for r in range(row, row + self.SIZE):
                    for c in range(col, col + self.SIZE):
                        if board[r][c] == board[i][j] and (r != i or c != j):
                            return False

        return True


class Solution:
    """
    Time:   O(n^2)
    Memory: O(n^2)
    """

    SIZE = 3
    EMPTY = '.'

    def isValidSudoku(self, board: List[List[str]]) -> bool:
        seen = (
            [(val, r), (c, val), (r // self.SIZE, c // self.SIZE, val)]
            for r, row in enumerate(board)
            for c, val in enumerate(row)
            if val != self.EMPTY
        )
        seen = sum(seen, [])
        return len(seen) == len(set(seen))

File: Java/Valid Sudoku.java
public class Solution {

    /**
     * Time:   O(n^3)
     * Memory: O(1)
     */
    public boolean isValidSudoku(char[][] board) {
        for (int i = 0; i < SIZE * SIZE; ++i) {
            for (int j = 0; j < SIZE * SIZE; ++j) {
                if (board[i][j] == EMPTY)
                    continue;

                for (int k = 0; k < SIZE * SIZE; ++k) {
                    if (board[i][k] == board[i][j] && k != j)
                        return false;
                    if (board[k][j] == board[i][j] && k != i)
                        return false;
                }

                int row = i - i % SIZE;
                int col = j - j % SIZE;


                for (int r = row; r < row + SIZE; ++r)
                    for (int c = col; c < col + SIZE; ++c)
                        if (board[r][c] == board[i][j] && (r != i || c != j))
                            return false;
            }
        }

        return true;
    }

    private static final int SIZE = 3;
    private static final char EMPTY = '.';
}
